### PR TITLE
Improve component accessibility

### LINKED ISSUE #5: auro-dialog: rapidly closing and opening the modal does not show the modal
## Describe the bug
If I close the modal and immediately open it again, the screen darkens but the modal is not displayed.

dialog

## To Reproduce
1. Open the dialog
2. Close the dialog and immediately open it again
3. The screen darkens but the content is not displayed

## Expected behavior
The content is always displayed when the modal is open.

## Desktop (please complete the following information):
 - OS: Windows 10
 - Browser: latest Firefox

## Additional context
I think it's related to these lines: https://github.com/AlaskaAirlines/auro-dialog/blob/d0eaee489248c929c5d8622e73337302c5d27816/src/auro-dialog.js#L113-L131

Why is a timeout needed here?

### LINKED ISSUE #7: auro-dialog: a11y - escape key does not close dialog
## Describe the bug
Per [WAI-ARIA authoring practices 1.1](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal), the escape key should close the dialog.

## To Reproduce
1. Open the dialog
2. Hit the escape key
3. Nothing happens

## Expected behavior
The escape key closes the dialog.

## Additional context
I assume this should only happen in a non-blocking context, i.e. when the close button is shown.

### LINKED ISSUE #8: auro-dialog: a11y - focus is not trapped within the dialog
## Describe the bug
Per [WAI-ARIA authoring practices 1.1](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal), focus should be trapped within the dialog.

> When a dialog opens, focus moves to an element inside the dialog. See notes below regarding initial focus placement.
>
> Tab:
> - Moves focus to the next tabbable element inside the dialog.
> - If focus is on the last tabbable element inside the dialog, moves focus to the first tabbable element inside the dialog.
>
> Shift + Tab:
>  - Moves focus to the previous tabbable element inside the dialog.
>  - If focus is on the first tabbable element inside the dialog, moves focus to the last tabbable element inside the dialog.

> When a dialog opens, focus placement depends on the nature and size of the content.
> 
> -  In all circumstances, focus moves to an element contained in the dialog.
> -  Unless a condition where doing otherwise is advisable, focus is initially set on the first focusable element.
> -  If content is large enough that focusing the first interactive element could cause the beginning of content to scroll out of view, it is advisable to add tabindex="-1" to a static element at the top of the dialog, such as the dialog title or first paragraph, and initially focus that element.
> -  If a dialog contains the final step in a process that is not easily reversible, such as deleting data or completing a financial transaction, it may be advisable to set focus on the least destructive action, especially if undoing the action is difficult or impossible. The Alert Dialog Pattern is often employed in such circumstances.
> -  If a dialog is limited to interactions that either provide additional information or continue processing, it may be advisable to set focus to the element that is likely to be most frequently used, such as an "OK" or "Continue" button.

## To Reproduce
1. Open the dialog with the keyboard. Notice that focus is not moved inside the dialog.
2. Tab through the dialog. Notice that you can escape the dialog.

## Expected behavior
When I open the dialog with the keyboard, my focus is moved inside the dialog. When I continue tabbing within the dialog, I am unable to escape the dialog.

On opening the dialog, focus should likely be moved to either the close button (if non-blocking) or the button that dismisses the dialog (if blocking).

## Additional context
This is important to give keyboard users the same experience as those navigating with a mouse. Mouse users are unable to trigger any elements outside the dialog, so keyboard users should be unable to as well. Currently, keyboard users can tab to any element on the page and activate it, even when the dialog is open.

### LINKED ISSUE #9: auro-dialog: a11y - screen readers can escape the active dialog
## Describe the bug
When the dialog is open, screen readers can still navigate to content outside of the dialog. Per the [WAI-ARIA authoring practices](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal) (emphasis mine):

> A dialog is a window overlaid on either the primary window or another dialog window. Windows under a modal dialog are inert. That is, _users cannot interact with content outside an active dialog window._

## To Reproduce
1. Open the dialog while using a screen reader
2. Attempt to navigate out of the screen reader (e.g. to next button or form field)
3. Notice you are able to escape the modal

## Expected behavior
Screen reader users can only interact and navigate to content within the modal.

## Desktop (please complete the following information):
 - OS: Windows 10
 - Browser: Firefox w/ NVDA

## Additional context
This is a separate issue from AlaskaAirlines/auro-interruption#8. Trapping keyboard focus does not necessarily trap a screen reader's virtual cursor. Setting `aria-modal="true"` could fix the issue, though there is [an open WebKit issue](https://bugs.webkit.org/show_bug.cgi?id=174667) where setting aria-modal prevents VoiceOver from reading static text within the dialog (not good). 

Another solution is [setting aria-hidden on content outside the dialog](https://developer.paciellogroup.com/blog/2018/06/the-current-state-of-modal-dialog-accessibility/). However we solve this, we will need to thoroughly test with desktop and mobile screen readers to make sure it is behaving as expected.

### LINKED ISSUE #10: auro-dialog: a11y - focus not returned to previously active element
## Describe the bug
When I close the modal using the keyboard, my focus is lost. Ideally, it should be returned to the previously-active element (e.g. the button that opened the modal). Per the WAI-ARIA Authoring Practices:

> When a dialog closes, focus returns to the element that invoked the dialog unless either: 
> - The invoking element no longer exists. Then, focus is set on another element that provides logical work flow.
> - The work flow design includes the following conditions that can occasionally make focusing a different element a more logical choice: 
>   1. It is very unlikely users need to immediately re-invoke the dialog.
>   2. The task completed in the dialog is directly related to a subsequent step in the work flow.
> 
> For example, a grid has an associated toolbar with a button for adding rows. the Add Rows button opens a dialog that prompts for the number of rows. After the dialog closes, focus is placed in the first cell of the first new row.

## To Reproduce
1. Open the dialog
2. Tab to the close button and trigger it
3. Keyboard focus is lost. At time of writing it remains on the close button (because of AlaskaAirlines/auro-dialog#6), but that is no longer visible on the page.

## Expected behavior
My focus is returned to the element that triggered the dialog.

## Desktop (please complete the following information):
 - OS: Windows 10
 - Browser: Firefox

## Additional context
Without implementing this, keyboard users will be confused as to where their focus went. They may think that they have to start tabbing from the start of the page.

### LINKED ISSUE #14: auro-dialog: address missing tests
# General Support Request
Test coverage is under 80% in some cases. 

```
=============================== Coverage summary ===============================
Statements   : 72.97% ( 27/37 )
Branches     : 92.31% ( 12/13 )
Functions    : 75% ( 9/12 )
Lines        : 72.97% ( 27/37 )
================================================================================
```

## Support request
Update missing test scenarios

### PR DESCRIPTION
**Pull request by geoffrich**
# Alaska Airlines Pull Request

This fixes multiple accessibility issues with the dialog.

fixes #7, fixes #10, fixes #14, fixes #5

Partially handles #8 and #9, though there are issues with focus escaping to other custom elements. These issues will need to be investigated further.

## Summary:
- Implement custom getter/setter to run accessibility enhancements (outlined below) when this.open changes
- Focus dialog on open and return focus to triggering element on close (#10)
- Add event listener to window to close dialog when ESC pressed, unless dialog is a modal (#7)
- Add aria-hidden and inert to outside elements when modal is opened to trap focus (#8, #9). This uses Google's [inert polyfill](https://github.com/GoogleChrome/inert-polyfill)
- Use role="dialog" instead of dialog element due to [screen reader issues with the dialog element](https://www.scottohara.me/blog/2019/03/05/open-dialog.html)
- Refactor transition to fix bug with rapidly opening and closing the dialog (#5). Transitioning visibility: hidden instead of using timeouts fixed the bug and the transitions appear the same as before. As a bonus, the code is cleaner too.
- Code cleanup and refactoring to better describe use of internal functions
- Increase code coverage to greater than 90% (#14)

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [x] Revision of an existing capability

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

**Comment by geoffrich** [comment]
The bulk of the code changes are done for this. Still need to do a little more testing with different browsers/screen readers.

**Comment by geoffrich** [comment]
Updated code coverage:
```
=============================== Coverage summary ===============================
Statements   : 97.06% ( 66/68 )
Branches     : 91.67% ( 33/36 )
Functions    : 96.43% ( 27/28 )
Lines        : 96.97% ( 64/66 )
================================================================================
```

**Comment by geoffrich** [comment]
While testing the PR, @blackfalcon discovered that he could focus other Auro custom elements while the dialog was open (video below). #8 and #9 will stay open until this issue is resolved. This PR can be merged as is.


https://user-images.githubusercontent.com/4992896/104786982-50812b80-5743-11eb-9bd0-fb4da49a6fe7.mov

File: docs/api.md [hunk 1]
```diff
@@ -19,6 +19,12 @@
 | `open`        | `open`        | `Boolean` | Sets state of dialog to open                     |
 | `unformatted` | `unformatted` | `Boolean` | Unformatted dialog window, edge-to-edge fill for content |
 
+## Methods
```
Comment (blackfalcon) [inline docs/api.md:22]: Need to have docs updated to better explain the use case of this method.

Comment (blackfalcon) [inline docs/api.md:22]: Need to have docs updated to better explain the use case of this method.

**Review by blackfalcon — APPROVED** [review]

**Pull request merged**